cmd/tailscale/cli/up.go

```diff
@@ -76,6 +77,37 @@ var upArgs struct {
 	snat            bool
 	netfilterMode   string
 	authKey         string
+	hostname        string
+}
+
+// validateHostname checks that name is a valid non-empty domain name label
```

Comment:
This seems to conflate host name and DNS name.

We have separate fields for both in the structs passed around to/from control.

And this check is only for an explicit one. If I set my system hostname to any non-DNS-compliant gibberish, we'll send that as-is today.

Some users have emojis in their hostnames. (we going to Punycode it? :))

Reply:
Certainly it's much easier to allow arbitrary (bounded-length?) strings as hostnames :)

Reply:
If that's more in line with the role we want hostname to play, I'll do that and use this scheme at the `Node.Name` generation step instead.

Reply:
Should we at least limit it to 63 characters? Otherwise it will have to get truncated when forming the domain name, which the user probably doesn't want.

Reply:
Or perhaps I'm off base and that RFC is not enforced in practice?

Reply:
I assume the eventual plan is to have Control pick the DNS names and assign the node.Name (or renamed node.DNSName) field, preferring the most unique name possible, else using the admin-defined override from some admin panel. So if your hostname happens to be valid & unique, it's used, in which case this validation doesn't matter. 

In general, client-side validation doesn't matter much if you have to do it later server-side anyway.

That said, this is also fine for now, even if it's too strict for some cases.